compiler/src/iree/compiler/Dialect/Stream/Conversion/HALToStream/Patterns.cpp

```diff
@@ -12,6 +12,10 @@
 #include "iree/compiler/Dialect/Stream/IR/StreamOps.h"
 #include "mlir/Dialect/Arith/IR/Arith.h"
 
+namespace mlir::iree_compiler::IREE::Encoding {
+bool hasPackedStorageAttr(mlir::RankedTensorType);
+} // namespace mlir::iree_compiler::IREE::Encoding
+
```

Comment:
IWYU. It is okay to make HAL depend on Encoding. You can include the `EncodingTypes.h`. Is there a cyclic dependency issue?

Reply:
There are no dependency issues here, just that adding a header file (instead of adding a forward declaration) adds additional parsing and compilation time. 

Both are identical in functionalities, should we prefer header files instead in the future? @hanhanW

Reply:
Yes, we prefer header files. Forward declaration can hide dependency. We should be careful on encoding dialect dependency, so I'd suggest avoiding forward declaration in this case.

https://google.github.io/styleguide/cppguide.html#Forward_Declarations